packages/signers/src/sign-transaction.ts

```diff
@@ -0,0 +1,202 @@
+import { deduplicateSigners } from './deduplicate-signers';
+import { isTransactionModifyingSigner, TransactionModifyingSigner } from './transaction-modifying-signer';
+import { isTransactionPartialSigner, TransactionPartialSigner } from './transaction-partial-signer';
+import { isTransactionSendingSigner, TransactionSendingSigner } from './transaction-sending-signer';
+import { isTransactionSigner, TransactionSigner } from './transaction-signer';
+
+type CompilableTransactionWithSigners = CompilableTransaction<
+    TransactionVersion,
+    IInstructionWithSigners<string, TransactionSigner>
+> &
+    Partial<ITransactionWithSignatures>;
+
+/**
+ * Signs a transaction using any signers that may be stored in IAccountSignerMeta instruction accounts
+ * as well as any signers provided explicitly to this function.
+ * It will ignore TransactionSendingSigners since this function does not send the transaction.
+ */
+export async function partiallySignTransactionWithSigners<
+    TTransaction extends CompilableTransactionWithSigners = CompilableTransactionWithSigners
+>(
+    transaction: TTransaction,
+    signers: readonly TransactionSigner[] = []
+): Promise<TTransaction & ITransactionWithSignatures> {
+    const { partialSigners, modifyingSigners } = categorizeTransactionSigners(
+        deduplicateSigners([...getSignersFromTransaction(transaction).filter(isTransactionSigner), ...signers]),
+        false
+    );
+
+    return signModifyingAndPartialTransactionSigners(transaction, modifyingSigners, partialSigners);
+}
+
+/**
+ * Signs a transaction using any signers that may be stored in IAccountSignerMeta instruction accounts
+ * as well as any signers provided explicitly to this function.
+ * It will assert that the transaction is fully signed before returning.
+ * It will ignore TransactionSendingSigners since this function does not send the transaction.
+ */
+export async function signTransactionWithSigners<
+    TTransaction extends CompilableTransactionWithSigners = CompilableTransactionWithSigners
+>(
+    transaction: TTransaction,
+    signers: readonly TransactionSigner[] = []
+): Promise<TTransaction & IFullySignedTransaction> {
+    const signedTransaction = await partiallySignTransactionWithSigners(transaction, signers);
+    assertTransactionIsFullySigned(signedTransaction);
+    return signedTransaction;
+}
+
+/**
+ * Signs and sends a transaction using any signers that may be stored in IAccountSignerMeta
+ * instruction accounts as well as any signers provided explicitly to this function.
+ * It will identify a single TransactionSendingSigners to use for sending the transaction, if any.
+ * Otherwise, it will send the transaction using the provided fallbackSender.
+ */
+export async function signAndSendTransactionWithSigners<
+    TTransaction extends CompilableTransactionWithSigners = CompilableTransactionWithSigners
+>(
+    transaction: TTransaction,
+    signers: readonly TransactionSigner[] = [],
```

Comment:
As per the documentation, wasn't this supposed to be optional?

Reply:
It is. See the ` = []` default value at the end. 🙂

Reply:
Ah, missed that!